File: packages/detached_membrane_sdk/detached_membrane_sdk/pim0_emit.py
"""Emit thin PIM0 envelopes from detached membrane proposals."""

from __future__ import annotations

from typing import Any

from .envelope_adapter import adapt_event_envelope


def emit_pim0_from_proposal(
    *,
    packet: dict[str, Any],
    source_repo: str = "agents-for-ollama",
    policy_version: str = "detached-membrane-policy-v1",
    envelope_id: str | None = None,
) -> dict[str, Any]:
    """Create a thin PIM0 envelope reference for a CAS proposal packet."""
    return_id = packet.get("return_id", "unknown_return")
    source_packet_id = packet.get("source_packet_id", "unknown_source")
    envelope = envelope_id or f"evt_dm_{return_id}"

    return adapt_event_envelope(
        source_repo=source_repo,
        source_object_type="CASReturnPacket",
        source_schema_ref="cas-return-0_1",
        source_object_ref=return_id,
        source_hash=f"b3:{return_id}",
        authority_class="presentation_only",
        lineage_ref=source_packet_id,
        policy_version=policy_version,
        mutation_allowed=False,
        receipt_ref=packet.get("ztna", {}).get("policy_decision_ref"),
        envelope_id=envelope,
    )

File: packages/detached-membrane-governance/src/executor/pim0-emit.ts
const PIM0_ENVELOPE_VERSION = "PIM0_EVENT_ENVELOPE_V1";
const VALID_AUTHORITY_CLASSES = new Set([
  "advisory",
  "proposal",
  "governed",
  "canonical",
  "presentation_only",
]);

export function adaptEventEnvelope(input: {
  sourceRepo: string;
  sourceObjectType: string;
  sourceSchemaRef: string;
  sourceObjectRef: string;
  sourceHash: string;
  authorityClass: string;
  lineageRef: string;
  policyVersion: string;
  mutationAllowed: boolean;
  receiptRef: string | null;
  envelopeId: string;
  observedAt?: string;
  now?: Date;
}): Record<string, unknown> {
  const required: Record<string, string> = {
    source_repo: input.sourceRepo,
    source_object_type: input.sourceObjectType,
    source_schema_ref: input.sourceSchemaRef,
    source_object_ref: input.sourceObjectRef,
    source_hash: input.sourceHash,
    lineage_ref: input.lineageRef,
    policy_version: input.policyVersion,
    envelope_id: input.envelopeId,
  };
  for (const [key, value] of Object.entries(required)) {
    if (!value.trim()) throw new Error(`${key} must be non-empty`);
  }
  if (!VALID_AUTHORITY_CLASSES.has(input.authorityClass)) {
    throw new Error(`authority_class must be one of: ${[...VALID_AUTHORITY_CLASSES].sort().join(", ")}`);
  }

  return {
    envelope_version: PIM0_ENVELOPE_VERSION,
    envelope_id: input.envelopeId,
    source_repo: input.sourceRepo,
    source_object_type: input.sourceObjectType,
    source_schema_ref: input.sourceSchemaRef,
    source_object_ref: input.sourceObjectRef,
    source_hash: input.sourceHash,
    authority_class: input.authorityClass,
    lineage_ref: input.lineageRef,
    policy_version: input.policyVersion,
    mutation_allowed: Boolean(input.mutationAllowed),
    receipt_ref: input.receiptRef,
    observed_at: input.observedAt ?? (input.now ?? new Date()).toISOString(),
  };
}

export function emitPim0FromProposal(
  packet: Record<string, unknown>,
  options?: { envelopeId?: string; now?: Date },
): Record<string, unknown> {
  const returnId = String(packet.return_id ?? "unknown_return");
  const sourcePacketId = String(packet.source_packet_id ?? "unknown_source");
  const envelopeId = options?.envelopeId ?? `evt_dm_${returnId}`;
  const ztna = (packet.ztna ?? {}) as Record<string, unknown>;

  return adaptEventEnvelope({
    sourceRepo: "agents-for-ollama",
    sourceObjectType: "CASReturnPacket",
    sourceSchemaRef: "cas-return-0_1",
    sourceObjectRef: returnId,
    sourceHash: `b3:${returnId}`,
    authorityClass: "presentation_only",
    lineageRef: sourcePacketId,
    policyVersion: "detached-membrane-policy-v1",
    mutationAllowed: false,
    receiptRef: (ztna.policy_decision_ref as string | undefined) ?? null,
    envelopeId,
    now: options?.now,
  });
}
